lstchain/reco/reconstructor.py

```diff
@@ -0,0 +1,944 @@
+from abc import abstractmethod, ABC
+import inspect
+import logging
+
+from iminuit import Minuit
+import numpy as np
+import numexpr as ne
+import matplotlib.pyplot as plt
+from copy import copy
+import astropy.units as u
+
+from ctapipe.reco.reco_algorithms import Reconstructor
+from lstchain.io.lstcontainers import DL1ParametersContainer
+from lstchain.image.pdf import log_gaussian, log_asygaussian2d
+from lstchain.visualization.camera import display_array_camera
+
+logger = logging.getLogger(__name__)
+
+
+class DL0Fitter(ABC):
+    """
+        Base class for the extraction of DL1 parameters from R1 events
+        using a log likelihood minimisation method.
+
+    """
+
+    def __init__(self, waveform, error, sigma_s, geometry, dt, time_shift, n_samples,
```

Comment:
Same comment as above. A fitter should be instantiated once per run in the script and only take options in the `__init__` method. Then it should take an event in `__call__` taking all options that change per event from there.

See e.g. how the HillasReconstructor in ctapipe works:

https://github.com/cta-observatory/ctapipe/blob/bc78946d752680189ffc49c1f848b6765b8dc6c4/ctapipe/reco/hillas_reconstructor.py#L153-L180